pydantic/schema.py

```diff
@@ -496,7 +498,10 @@ def field_type_schema(
 
     # check field type to avoid repeated calls to the same __modify_schema__ method
     if field.type_ != field.outer_type_:
-        modify_schema = getattr(field.outer_type_, '__modify_schema__', None)
+        if field.shape == SHAPE_GENERIC:
+            modify_schema = getattr(field.type_, '__modify_schema__', None)
+        else:
+            modify_schema = getattr(field.outer_type_, '__modify_schema__', None)
```

Comment:
suggestion for DRYness

```suggestion
        if field.shape == SHAPE_GENERIC:
            field_type = field.type_
        else:
            field_type = field.outer_type_
        modify_schema = getattr(field_type, '__modify_schema__', None)
```